Add test to ensure no index out of bounds panic (#135474)

reddevilmidzy opened the pull request:
Adds test for #135474

reddevilmidzy committed:
Add test to ensure no index out of bounds panic (#135474)

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13861838061/job/38792045567) [(plain)](https://github.com/rust-lang/rust/commit/d212e9f2162857ba81129d8dd1f48fd0ae18e306/checks/38792045567/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 3.005 Building wheels for collected packages: reuse
#12 3.006   Building wheel for reuse (pyproject.toml): started
#12 3.234   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.235   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.235   Stored in directory: /tmp/pip-ephem-wheel-cache-cfihiqo2/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.238 Successfully built reuse
#12 3.238 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.666 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.667 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.270 Collecting virtualenv
#12 4.329   Downloading virtualenv-20.29.3-py3-none-any.whl (4.3 MB)
#12 4.483      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 28.3 MB/s eta 0:00:00
#12 4.543 Collecting platformdirs<5,>=3.9.1
#12 4.552   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.580 Collecting distlib<1,>=0.3.7
#12 4.588   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.600      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 47.9 MB/s eta 0:00:00
#12 4.644 Collecting filelock<4,>=3.12.2
#12 4.652   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.739 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.934 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.6 virtualenv-20.29.3
#12 4.935 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 5.0s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      124864 kB
DirectMap2M:     8263680 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
WARN: `rust.debug-assertions = true` will prevent downloading CI rustc as alt CI rustc is not currently built with debug assertions.
---
fmt check
fmt: checked 5905 files
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: /checkout/tests/ui/impl-sized.rs: too many trailing newlines (2)
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:55
  local time: Fri Mar 14 17:24:55 UTC 2025
  network time: Fri, 14 Mar 2025 17:24:55 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

compiler-errors reviewed (changes requested):
This test needs a better name and should be placed into a *subdirectory* of `tests/ui` corresponding to what the test is exercising, rather than the top-level.

Specifically, I think you should look back at the original issue and/or the PR that fixed the issue to understand what was wrong with the code in the first place and/or why the issue was fixed, so that this can be named more accurately.

reddevilmidzy commented:
Thanks for the feedback! I've moved the test to the `tests/ui/fn` directory.  I also reviewed the original PR (#136497) to better understand the context and ensure the test is more accurately placed.

Additionally, I've renamed the test from `impl-sized` to `trait-fn-generic-mismatch`

Let me know if you have any further suggestions!

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13870946473/job/38817488490) [(plain)](https://github.com/rust-lang/rust/commit/cd243846263032e93adc42f3f542ca2d2d705ef7/checks/38817488490/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.795 Building wheels for collected packages: reuse
#12 2.797   Building wheel for reuse (pyproject.toml): started
#12 3.010   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.011   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.011   Stored in directory: /tmp/pip-ephem-wheel-cache-9klra3oi/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.013 Successfully built reuse
#12 3.014 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.417 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.418 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 3.961 Collecting virtualenv
#12 4.003   Downloading virtualenv-20.29.3-py3-none-any.whl (4.3 MB)
#12 4.126      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 35.8 MB/s eta 0:00:00
#12 4.166 Collecting distlib<1,>=0.3.7
#12 4.170   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.179      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 70.8 MB/s eta 0:00:00
#12 4.212 Collecting platformdirs<5,>=3.9.1
#12 4.217   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.257 Collecting filelock<4,>=3.12.2
#12 4.261   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.343 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.528 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.6 virtualenv-20.29.3
#12 4.528 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.6s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      128960 kB
DirectMap2M:     7211008 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
WARN: `rust.debug-assertions = true` will prevent downloading CI rustc as alt CI rustc is not currently built with debug assertions.
---
fmt check
fmt: checked 5905 files
tidy check
tidy: Skipping binary file check, read-only filesystem
tidy error: /checkout/tests/ui/fn/trait-fn-generic-mismatch.rs: too many trailing newlines (2)
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:48
  local time: Sat Mar 15 08:10:06 UTC 2025
  network time: Sat, 15 Mar 2025 08:10:06 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

compiler-errors commented:
Please squash this into one commit. there doesn't need to be a commit to move a file or to remove whitespace, for example. It's just one file after all :) Then I can approve.

@rustbot author

reddevilmidzy commented:
Squashed the commits into one as requested. Thanks for the review! 😊

@rustbot ready

compiler-errors commented:
@bors r+ rollup

bors commented:
:pushpin: Commit b52330136837aadf977b1f188cb5d60a264900d6 has been approved by `compiler-errors`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=compiler-errors b52330136837aadf977b1f188cb5d60a264900d6 -->
<!-- homu: {"type":"Approved","sha":"b52330136837aadf977b1f188cb5d60a264900d6","approver":"compiler-errors","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
